File: readability.js
// usage:
//    cat fixtures/ciokorea.html | node readability.js https://www.ciokorea.com/news/234834
var { Readability } = require('@mozilla/readability');
var { JSDOM } = require('jsdom');

var fs = require("fs");
var stdinBuffer = fs.readFileSync(0); // STDIN_FILENO = 0

var doc = new JSDOM(stdinBuffer.toString(), {
  url: process.argv[2],
});

let reader = new Readability(doc.window.document);
let article = reader.parse();

//
// title
// byline
// dir
// lang
// content
// textContent
// length
// siteName
console.log(JSON.stringify(article))

File: readability.go
package reader

import (
	"context"
	"encoding/json"
	"io"

	"github.com/whitekid/goxp"
)

type Article struct {
	Title       string
	Byline      string
	Dir         string
	Content     string
	TextContent string
	Length      int
	Excerpt     string
	SiteName    string
}

func readableArticle(ctx context.Context, r io.Reader, url string) (*Article, error) {
	var article Article
	var parseErr error

	if err := goxp.Exec("node", "readability.js", url).
		Pipe(
			func(stdin io.Writer) { io.Copy(stdin, r) },
			func(stdout io.Reader) { parseErr = json.NewDecoder(stdout).Decode(&article) },
			nil).
		Do(ctx); err != nil {
		return nil, err
	}

	if parseErr != nil {
		return nil, parseErr
	}

	return &article, nil
}
